Use OOP forcefields

a-ws-m opened the pull request:
This implements several changes to make it easier for end-users to implement their own forcefields. Most of these changes revolve around the `Forcefield` class, which stores the path dictionaries that were previously separate entities. To use a non-default forcefield, make a new `Forcefield` instance and then pass it to `Simulation` using the `ff_class` argument. This argument is optional; passing a string value for `forcefield` will still work, it's simply converted behind the scenes. I also tidied up a few things, which may be worth discussing:

* The `system.top` and `system_octwet.top` files have been changed to be compatible with Python's built-in [Template](https://docs.python.org/3/library/string.html?highlight=template#string.Template) functionality. This makes the code a bit easier to understand.
* Water models are now attributes of the individual forcefields, and are read directly from the corresponding `watermodels.dat` file. As I understand it, GROMACS_WATER_MODELS is a master list of all water models, some of which are only available for OPLS-AA. If this is correct, it might be good to phase this out. I have changed the core functions that used this: `get_solvent_model` and `get_solvent_identifier`. The only other places in the code where `GROMACS_WATER_MODELS` is used is in the tests and in `get_water_model`, a function which is also unused in the code base except in tests.
* I have changed one test to check for equality rather than equivalence (`is` -> `==`).

I'll write some example code that demonstrates how to use a `Forcefield` next week. In the meantime, I'm sharing this PR so that the changes can be discussed.

a-ws-m commented:
@orbeckst here's the PR you asked me to ping you about!

orbeckst reviewed (changes requested):
Thank you for the PR. In general I like your approach
- consolidate in `Forcefield` class
- make the templates real templates

I have two initial requests:

1. The original intention with the templates was that they should in principle be legal input form GROMACS. However, I like explicit templating better (also, with 10+ experience since I wrote the original code...) so could you please rename the files from `*.top` to `*.top.template` (or some other appropriate suffix... `*.top.in`?) to make clear that these are now *template files*?

2. You seem to be using an autoformatter which makes reading the diff difficult. Can you please either undo the formatting or submit a separate PR _only with the formatting changes on the files that you touched_. We can then merge that PR and then we can merge the main into this PR (or you can rebase) and we have a clean diff.

orbeckst commented:
You gave me the motivation to try out black on a project so I opened #271. Let's do PR #271 first and then come back here.

orbeckst commented:
@a-ws-m now that PR #269 is merged, could you please rebase against current develop so that we hopefully end up with a nice, clean diff? Thanks!

a-ws-m commented:
I've made a few important changes:

* The `solvation` errors arose because Gromacs expects a `.top` file suffix, but the copies of the templates were saved with identical names (i.e., including `.template`). I have added a few lines to ensure that the topology file _does_ have the correct suffix.
* `forcefields.get_solvent_identifier()` is now consistent with `get_solvent_model()` and raises an error when no such identifier exists for the given force field.
* I removed the `python.system_packages` key from the read the docs configuration as it was [just deprecated ](https://github.com/readthedocs/readthedocs.org/commit/bcac96225d3d613c3c19b41b5f33c80b4f31a120) and was [causing build errors](https://readthedocs.org/projects/mdpow/builds/21756823/).

a-ws-m commented:
Regarding [the error](https://github.com/Becksteinlab/MDPOW/actions/runs/6027967217/job/16354289437?pr=267#step:9:603) in the FEP tests, I'm unable to reproduce it on my machine and I think it's an issue with the MD workflow rather than any Python code that I've changed here.

orbeckst commented:
> Regarding [the error](https://github.com/Becksteinlab/MDPOW/actions/runs/6027967217/job/16354289437?pr=267#step:9:603) in the FEP tests, I'm unable to reproduce it on my machine and I think it's an issue with the MD workflow rather than any Python code that I've changed here.

There's nothing you can do. It's a stochastic failure: if the simulation ends randomly reaches a point where the solute extends more than the nb cutoff, it triggers a known issue with how GROMACS calculates exclusions and modern versions of GROMACS stop the simulation (old versions just kept going with potentially questionable results).

Perhaps we can try an set a random seed that leads to a known working testrun... or we eventually replace it with a less iffy system.

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -55,35 +57,25 @@
 
 logger = logging.getLogger("mdpow.forecefields")
 
+HERE = Path(__file__).parent
+TOP_DIR = HERE / "top"
+
```
remove, see comments below, you only need
```python
from . import config
```

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
-                doi: 10.1021/acsomega.8b03132"""
+NEW_OCTANOL_DESC = "Zangi R (2018) Refinement of the OPLSAA force-field for liquid alcohols.; ACS Omega 3(12):18089-18099.  doi: 10.1021/acsomega.8b03132"
+
+OPLS_AA_FF_DIR = TOP_DIR / "oplsaa.ff"
```
Instead of the HERE approach, use
```suggestion
OPLS_AA_FF_DIR = Path(config.topfiles["oplsaa.ff"])
```
because in this way we control all access to bundled files from `mdpow.config`.

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
-        identifier="toluene", itp="1tol.itp", coordinates="1tol_oplsaa.gro"
-    ),
-}
+CHARMM_FF_DIR = TOP_DIR / "charmm36-mar2019.ff"
```
```suggestion
CHARMM_FF_DIR = Path(config.topfiles["charmm36-mar2019.ff"])
```

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
-        identifier="toluene", itp="1tol.itp", coordinates="1tol_charmm.gro"
-    ),
-}
+AMBER_FF_DIR = TOP_DIR / "amber99sb.ff"
```
```suggestion
CHARMM_FF_DIR = Path(config.topfiles["amber99sb.ff"])
```

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
+    of these models is a bit convoluted and it is best to obtain the desired
+    water model in these two steps::
+
+      identifier = get_solvent_identifier("water", model="tip3p") model =
```
line break before model !

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
+    :Returns: An identifier
     """
+    if isinstance(forcefield, str):
+        forcefield = _get_forcefield(forcefield)
```
Can you please test this line? It seems as if this would be something people use often.

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
             raise ValueError(msg)
 
 
-def get_ff_paths(forcefield="OPLS-AA"):
```
nice, I like removing whole functions — the new organization by FF makes more sense!

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
 
+@dataclass
+class Forcefield:
+    """Contains information about files corresponding to a forcefield."""
```
add a line
```reST

.. versionadded: 0.9.0
```

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
+    :Raises ValueError: If there is no identifier found for the combination.
 
-    :Returns: Either an identifier or ``None``
+    :Returns: An identifier
```
add a line
```reST

.. versionchanged:: 0.9.0
   Raises :exc:`ValueError` instead of returning ``None``.
```

Also note this change of behavior in CHANGES.

orbeckst commented on mdpow/forcefields.py:
```diff
@@ -180,163 +229,228 @@ def get_water_model(watermodel=DEFAULT_WATER_MODEL):
 
-    If identifier is "water" then the :data:`DEFAULT_WATER_MODEL` is assumed.
+    If identifier is "water" then the default water model for the :class:`Forcefield` is assumed.
+
```
Add a
```reST
.. versionchanged:: 0.9.0
   Function can now also accept a :class:`Forcefield` for the `forcefield argument.
```

orbeckst reviewed (changes requested):
This looks really good overall, I like the new `Forcefield` class that combines all information in a sensible data structure.

I have a bunch of comments — please address them.

Please also add entries to CHANGES. Some of the comments ask you to note changes and additions (e.g. reorganization of how force fields are described with the new `forcefields.Forcefield` class). Also note which template files have been renamed and are now true template files.

a-ws-m commented:
I've made those changes. The GitHub tests seem to be throwing some weird errors to do with cloning the repo.

orbeckst commented:
I restarted CI and all jobs are running. It was a temporary problem.

a-ws-m commented:
No problem! I'll work on those. I have actually been working on [an example](https://github.com/a-ws-m/MDPOW/blob/martini-example/doc/examples/martini/martini-benzene.ipynb), it's in the `martini-example` branch of my fork. I have some questions before it's ready that I'll raise on the Q&A.

orbeckst commented:
awesome!

a-ws-m commented:
I've merged the example using Martini, you can find it in `doc/examples/martini-example`. I imagine there's scope for improving the run parameters to get a more accurate value, and perhaps fixing the issue whereby the `nan` error for the coulomb contribution propagates and gives a `nan` value for the overall error. I'm not sure how to improve these, though -- I'd be happy to hear suggestions!

orbeckst commented:
Your [MARTINI/MDPOW example notebook](https://github.com/Becksteinlab/MDPOW/blob/a80ecb3a4b3ff38e2f0fc65f31c84dd78fd39536/doc/examples/martini/martini-benzene.ipynb) is fantastic!

(I also noted that some of the GROMACS runs actually didn't run but just restarted from the last frame. However, it's nice to know that it's not an issue re-running the workflow as it won't destroy data.)

For nan value in the Coulomb calculation I am not sure where it comes from. I assume that there is a valid Coulomb leg in the FEP calculation because MARTINI benzene has partial charges, doesn't it?

One thing that's a bit odd is that alchemlyb and mdpow report different values for statistical inefficiency, for example alchemlyb states 15.84 and mdpow states 7.9840. 
```
mdpow.fep   : INFO     Performing statistical inefficiency analysis for window coulomb 0000
2023-09-04 22:41:16.809 | DEBUG    | alchemlyb.preprocessing.subsampling:statistical_inefficiency:520 - Running statistical inefficiency analysis.
2023-09-04 22:41:16.812 | DEBUG    | alchemlyb.preprocessing.subsampling:statistical_inefficiency:522 - Statistical inefficiency: 15.84.
2023-09-04 22:41:16.813 | DEBUG    | alchemlyb.preprocessing.subsampling:statistical_inefficiency:528 - Number of uncorrelated samples: 312.
mdpow.fep   : INFO     The statistical inefficiency value is 7.9840.
mdpow.fep   : INFO     The data are subsampled every 8 frames.
```
I think that's because we have to infer the actual value from what alchemlyb subsample in 
https://github.com/Becksteinlab/MDPOW/blob/6449879a9bc5fbc80210290eea84fd40ad164dff/mdpow/fep.py#L1192C20-L1192C20
and there's also a suspicious looking factor of 2 involved. (The whole question about getting the raw SI from alchemlyb was the topic of https://github.com/alchemistry/alchemlyb/issues/295 ). @VOD555 can you please check that the MDPOW output is consistent with the one from alchemlyb? I assume that other users that read the log output will have the same question as I

VOD555 commented:
@orbeckst I've checked the code. The logger in the mdpow outputs the correlation time (g/2) but not the statistical inefficiency (g). And as the one calculated in mdpow is based on the length output of the subsampled dataframe, there is small difference between the mdpow one and alchemlyb one.

As currently alchemlyb has a logger output for the SI, I think we can remove the mdpow one and add print out the number of frames of the subsampled dataframe.

VOD555 commented:
Open an issue #275

a-ws-m commented:
Thank you! The Gromacs output is for a restart because it was the quickest way to re-generate most of the useful information after clearing the cell outputs. I'll work on those other points today.

orbeckst commented on doc/examples/martini/run.mdp:
```diff
@@ -0,0 +1,108 @@
+
+; Which intermediate state do we start with?  Doesn't really matter, it leaves soon
+;-------
+init-lambda-state        = sedstate
```
*sedstate* is a weird value?

orbeckst reviewed (approved):
I am very happy with your contribution. It's a really neat re-organization that's also backwards compatible. 

Your example notebook is also really good. Nice work!!

orbeckst commented:
Thank you very much @a-ws-m , great contribution!!
